libs/testing/nest/src/lib/testServer.ts

```diff
@@ -0,0 +1,69 @@
+import {
+  Type,
+  ValidationPipe,
+  ExecutionContext,
+  INestApplication,
+} from '@nestjs/common'
+import { Test } from '@nestjs/testing'
+import { TestingModuleBuilder } from '@nestjs/testing/testing-module.builder'
+
+import {
+  IdsAuthGuard,
+  IdsUserGuard,
+  ScopesGuard,
+  getRequest,
+  User,
+} from '@island.is/auth-nest-tools'
+import { InfraModule } from '@island.is/infra-nest-server'
+
+export type TestServerOptions<AppModule> = {
+  appModule: AppModule
+  override?: (builder: TestingModuleBuilder) => TestingModuleBuilder
+  currentUser?: User
+  hooks?: ((app: INestApplication) => Promise<INestApplication>)[]
+}
+
+export const testServer = async <AppModule>({
+  appModule,
+  currentUser,
+  hooks = [],
+  override,
+}: TestServerOptions<AppModule>): Promise<INestApplication> => {
+  let builder = Test.createTestingModule({
+    imports: [InfraModule.forRoot(appModule as any)],
+  })
+
+  if (currentUser) {
+    builder = builder
+      .overrideGuard(IdsAuthGuard)
+      .useValue({ canActivate: () => true })
+      .overrideGuard(IdsUserGuard)
+      .useValue({
+        canActivate: (context: ExecutionContext) => {
+          const request = getRequest(context)
+          request.user = currentUser
+          return true
+        },
+      })
+      .overrideGuard(ScopesGuard)
+      .useValue({ canActivate: () => true })
+  }
+
+  if (override) {
+    builder = override(builder)
+  }
+
+  const moduleRef = await builder.compile()
+  let app: INestApplication = await moduleRef
+    .createNestApplication()
+    .useGlobalPipes(
+      new ValidationPipe({ whitelist: true, forbidNonWhitelisted: true }),
+    )
+    .init()
+
+  hooks.forEach(async (hook) => {
+    app = await hook(app)
+  })
```

Comment:
Not convinced about this API but it's a good start and should be easy to refactor when more use cases are discovered.

Reply:
I've extended the API, making it more powerful. I know I'm biased, but I find it nice being able to opt-in based on how I want to test stuff. Current implementation is like:

### Suitable for unit-tests
```typescript
app = await testServer({
  appModule: AppModule,
})
```

### Suitable for integration/functional-tests
```typescript
app = await testServer({
  appModule: AppModule,
  hooks: [
    useAuth({ currentUser }), // can skip if we want to test unauthorized behaviour
    useDatabase({ type: 'sqlite', provider: SequelizeConfigService }),
  ],
})
```
```typescript
app = await testServer({
  appModule: AppModule,
  hooks: [
    useAuth({ currentUser }),
    useDatabase({ type: 'postgres', provider: SequelizeConfigService }),
  ],
})
```

Reply:
Awesome, looks great!